### PR TITLE
*: refine CharsetAndCollation

### LINKED ISSUE #29560: refine CharsetAndCollation
## Enhancement
image

When we call CharsetAndCollation, c.charset and c.collation can't be nil. 
So the rest of code can't be reached. We better remove these codes.

### PR DESCRIPTION
**Pull request by wjhuang2016**
Signed-off-by: wjhuang2016 <huangwenjun1997@gmail.com>

### What problem does this PR solve?

Issue Number: close #29560

Problem Summary:
When we call CharsetAndCollation, c.charset and c.collation can't be nil. So the rest of code can't be reached. We better remove these codes.

### What is changed and how it works?
1. Refine CharsetAndCollation
2. Add default charset and collation into the inner session.
### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
None
```

### CODE REVIEW

**Review by xiongjiwei — APPROVED** [review]

**Review by zimulala — APPROVED** [review]
LGTM

File: expression/collation.go [hunk 1]
```diff
@@ -100,7 +90,7 @@ type CollationInfo interface {
 	SetRepertoire(r Repertoire)
 
 	// CharsetAndCollation ...
```
Comment (bb7133) [inline expression/collation.go:92]: Could you update the comments?

File: expression/collation.go [hunk 2]
```diff
@@ -68,17 +68,7 @@ func (c *collationInfo) SetCharsetAndCollation(chs, coll string) {
 	c.charset, c.collation = chs, coll
 }
 
-func (c *collationInfo) CharsetAndCollation(ctx sessionctx.Context) (string, string) {
-	if c.charset != "" || c.collation != "" {
-		return c.charset, c.collation
-	}
-
-	if ctx != nil && ctx.GetSessionVars() != nil {
```
Comment (bb7133) [inline expression/collation.go:76]: Why they are removed? Are you sure it is OK to make the change?

Comment (wjhuang2016) [reply expression/collation.go:76]: As I said in 'Problem summary':

When we call CharsetAndCollation, c.charset and c.collation can't be nil. So the rest of code can't be reached.

**Review by Reminiscent — APPROVED** [review]

**Review by bb7133 — APPROVED** [review]

**Pull request merged**